test/mpi/main.cc

```diff
@@ -383,34 +397,103 @@ void ProcessArgs(int argc, char** argv)
     }
 }
 
+#ifdef HAVE_CUDA
+void cuda_set_device()
+{
+    MPI_Comm local_comm;
+    int cuda_dev_count;
+    int local_rank;
+    MPI_Comm_split_type(MPI_COMM_WORLD, OMPI_COMM_TYPE_NODE, 0, MPI_INFO_NULL,
+                        &local_comm);
+    MPI_Comm_rank(local_comm, &local_rank);
+    CUDA_CHECK(cudaGetDeviceCount(&cuda_dev_count));
+    if(local_rank >= cuda_dev_count) {
```

Comment:
Sometimes maybe it worth setting device = (local_rank % cuda_dev_count). Maybe make --set-device option with paramter: 0 - don't set, 1 - set device=local_rank and error if local_rank >= cuda_dev_count and 2 - device = local_rank % cuda_dev_count.

Reply:
fixed